File: src/lib/db/writeInteraction.ts
/**
 * writeInteraction — records outreach events and inbound responses.
 *
 * Writes to:
 *   1. BigQuery hephae.interactions — permanent event log
 *   2. Firestore businesses/{slug}.crm — current CRM state update
 *
 * IMPORTANT: Only 'contact_form' and 'email_replied' are reliable responded signals.
 * 'email_opened' and 'report_link_clicked' are logged but do NOT flip crm.status.
 */

import { Timestamp } from 'firebase-admin/firestore';
import { db } from '@/lib/firebase';
import { bq, DATASET } from './bigquery';

export type InteractionEventType =
    | 'report_sent'           // outbound — initial report email
    | 'follow_up_sent'        // outbound — 2nd or 3rd touch
    | 'email_opened'          // inbound signal — NOT a reliable responded indicator
    | 'report_link_clicked'   // inbound signal — NOT a reliable responded indicator
    | 'contact_form'          // inbound — genuine response signal
    | 'email_replied';        // inbound — genuine response signal

const GENUINE_RESPONSE_EVENTS: InteractionEventType[] = ['contact_form', 'email_replied'];

export interface WriteInteractionOptions {
    businessSlug: string;
    zipCode?: string;
    eventType: InteractionEventType;
    contactEmail?: string;
    subject?: string;
    reportUrl?: string;
    outreachNumber?: number;  // 1, 2, or 3 — only for outbound events
}

export async function writeInteraction(opts: WriteInteractionOptions): Promise<void> {
    const { businessSlug, zipCode, eventType, contactEmail, subject, reportUrl, outreachNumber } = opts;
    const occurredAt = new Date();
    const interactionId = `${eventType}-${businessSlug}-${occurredAt.getTime()}`;
    const isGenuineResponse = GENUINE_RESPONSE_EVENTS.includes(eventType);

    // --- 1. Firestore CRM state update ---
    try {
        const isOutbound = eventType === 'report_sent' || eventType === 'follow_up_sent';
        const firestoreUpdate: Record<string, unknown> = {
            updatedAt: Timestamp.fromDate(occurredAt),
        };

        if (isOutbound && outreachNumber !== undefined) {
            firestoreUpdate['crm.outreachCount'] = outreachNumber;
            firestoreUpdate['crm.status'] = 'outreached';
            firestoreUpdate['crm.lastOutreachAt'] = Timestamp.fromDate(occurredAt);
            if (reportUrl) firestoreUpdate['crm.lastReportShared'] = reportUrl;
        }

        if (isGenuineResponse) {
            firestoreUpdate['crm.status'] = 'responded';
            firestoreUpdate['crm.respondedAt'] = Timestamp.fromDate(occurredAt);
        }

        await db.doc(`businesses/${businessSlug}`).set(firestoreUpdate, { merge: true });
    } catch (err) {
        console.error(`[DB] Firestore writeInteraction failed for ${businessSlug}:`, err);
    }

    // --- 2. BigQuery append (all events, including non-response signals) ---
    const row = {
        interaction_id: interactionId,
        occurred_at: occurredAt.toISOString(),
        business_slug: businessSlug,
        zip_code: zipCode ?? null,
        event_type: eventType,
        outreach_number: outreachNumber ?? null,
        contact_email: contactEmail ?? null,
        subject: subject ?? null,
        report_url: reportUrl ?? null,
        responded: isGenuineResponse,
    };

    bq.dataset(DATASET).table('interactions').insert([row], { skipInvalidRows: false })
        .catch(err => {
            console.error(`[DB] BQ interactions write failed for ${interactionId}:`, err?.errors ?? err);
        });
}

export async function archiveBusiness(
    businessSlug: string,
    reason: 'no_response' | 'not_interested' | 'converted'
): Promise<void> {
    try {
        await db.doc(`businesses/${businessSlug}`).set({
            'crm.status': 'archived',
            'crm.archivedAt': Timestamp.now(),
            'crm.archiveReason': reason,
            updatedAt: Timestamp.now(),
        }, { merge: true });
    } catch (err) {
        console.error(`[DB] archiveBusiness failed for ${businessSlug}:`, err);
    }
}

File: web/backend/lib/db/write_interaction.py
"""
writeInteraction — records outreach events and inbound responses.

Writes to:
  1. BigQuery hephae.interactions — permanent event log
  2. Firestore businesses/{slug}.crm — current CRM state update

IMPORTANT: Only 'contact_form' and 'email_replied' are reliable responded signals.
'email_opened' and 'report_link_clicked' are logged but do NOT flip crm.status.
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime, timezone
from typing import Optional

logger = logging.getLogger(__name__)

GENUINE_RESPONSE_EVENTS = {"contact_form", "email_replied"}


async def write_interaction(
    business_slug: str,
    event_type: str,
    zip_code: Optional[str] = None,
    contact_email: Optional[str] = None,
    subject: Optional[str] = None,
    report_url: Optional[str] = None,
    outreach_number: Optional[int] = None,
) -> None:
    """Record an interaction event to Firestore + BigQuery."""
    from backend.lib.firebase import db
    from backend.lib.bigquery import bq_insert

    occurred_at = datetime.now(timezone.utc)
    interaction_id = f"{event_type}-{business_slug}-{int(occurred_at.timestamp() * 1000)}"
    is_genuine_response = event_type in GENUINE_RESPONSE_EVENTS

    # --- 1. Firestore CRM state update ---
    # Use update() so dotted paths like crm.status are treated as nested fields.
    try:
        is_outbound = event_type in ("report_sent", "follow_up_sent")
        update_payload: dict[str, object] = {"updatedAt": occurred_at}
        has_update = False

        if is_outbound and outreach_number is not None:
            update_payload["crm.outreachCount"] = outreach_number
            update_payload["crm.status"] = "outreached"
            update_payload["crm.lastOutreachAt"] = occurred_at
            if report_url:
                update_payload["crm.lastReportShared"] = report_url
            has_update = True

        if is_genuine_response:
            update_payload["crm.status"] = "responded"
            update_payload["crm.respondedAt"] = occurred_at
            has_update = True

        if has_update:
            db.document(f"businesses/{business_slug}").update(update_payload)
    except Exception as err:
        # NOT_FOUND is fine — agent result may not be written yet
        if not (hasattr(err, "code") and callable(err.code) and err.code().value[0] == 5):
            logger.error(f"[DB] Firestore writeInteraction failed for {business_slug}: {err}")

    # --- 2. BigQuery append (all events) ---
    row = {
        "interaction_id": interaction_id,
        "occurred_at": occurred_at,
        "business_slug": business_slug,
        "zip_code": zip_code,
        "event_type": event_type,
        "outreach_number": outreach_number,
        "contact_email": contact_email,
        "subject": subject,
        "report_url": report_url,
        "responded": is_genuine_response,
    }

    asyncio.get_event_loop().run_in_executor(
        None,
        lambda: _bq_insert_sync(bq_insert, "interactions", row, interaction_id),
    )


async def archive_business(business_slug: str, reason: str) -> None:
    """Archive a business in Firestore CRM."""
    from backend.lib.firebase import db

    try:
        db.document(f"businesses/{business_slug}").set(
            {
                "crm.status": "archived",
                "crm.archivedAt": datetime.now(timezone.utc),
                "crm.archiveReason": reason,
                "updatedAt": datetime.now(timezone.utc),
            },
            merge=True,
        )
    except Exception as err:
        logger.error(f"[DB] archiveBusiness failed for {business_slug}: {err}")


def _bq_insert_sync(bq_insert_fn, table: str, row: dict, interaction_id: str) -> None:
    """Synchronous wrapper for fire-and-forget BQ insert."""
    import asyncio

    try:
        asyncio.run(bq_insert_fn(table, row))
    except Exception as err:
        logger.error(f"[DB] BQ interactions write failed for {interaction_id}: {err}")
